File: src/ai/memory.py
"""
EVOS AI - Memory System
Vector-based memory for storing and retrieving browsing history and knowledge
"""
import asyncio
import json
import os
from datetime import datetime
from typing import Dict, Any, List, Optional
from pathlib import Path
import sqlite3
import hashlib

from config import settings

# Try to import chromadb, fall back to simple SQLite if not available
try:
    import chromadb
    from chromadb.config import Settings as ChromaSettings
    CHROMADB_AVAILABLE = True
except ImportError:
    CHROMADB_AVAILABLE = False
    print("[Memory] ChromaDB not available, using SQLite fallback")


class MemoryStore:
    """
    Hybrid memory system combining:
    - SQLite for structured data (pages, entities, relationships)
    - ChromaDB for vector search (semantic similarity)
    """
    
    def __init__(self, db_path: str = None):
        self.db_path = db_path or os.path.join(settings.memory_db_path, "memory.db")
        self.vector_db = None
        self.collection = None
        self._initialized = False
        
    async def initialize(self) -> bool:
        """Initialize the memory system"""
        try:
            # Ensure directory exists
            os.makedirs(os.path.dirname(self.db_path), exist_ok=True)
            
            # Initialize SQLite
            self._init_sqlite()
            
            # Initialize ChromaDB if available
            if CHROMADB_AVAILABLE:
                self._init_chromadb()
            
            self._initialized = True
            print(f"[Memory] Initialized at {self.db_path}")
            return True
            
        except Exception as e:
            print(f"[Memory] Initialization failed: {e}")
            return False
    
    def _init_sqlite(self):
        """Initialize SQLite database"""
        conn = sqlite3.connect(self.db_path)
        cursor = conn.cursor()
        
        # Pages table
        cursor.execute('''
            CREATE TABLE IF NOT EXISTS pages (
                id TEXT PRIMARY KEY,
                url TEXT UNIQUE NOT NULL,
                title TEXT,
                domain TEXT,
                content TEXT,
                summary TEXT,
                first_visit TIMESTAMP,
                last_visit TIMESTAMP,
                visit_count INTEGER DEFAULT 1,
                importance_score REAL DEFAULT 0.5,
                tags TEXT,
                metadata TEXT
            )
        ''')
        
        # Entities table (people, products, concepts extracted from pages)
        cursor.execute('''
            CREATE TABLE IF NOT EXISTS entities (
                id TEXT PRIMARY KEY,
                name TEXT NOT NULL,
                type TEXT,
                description TEXT,
                first_seen TIMESTAMP,
                last_seen TIMESTAMP,
                occurrence_count INTEGER DEFAULT 1,
                metadata TEXT
            )
        ''')
        
        # Page-Entity relationships
        cursor.execute('''
            CREATE TABLE IF NOT EXISTS page_entities (
                page_id TEXT,
                entity_id TEXT,
                relevance_score REAL DEFAULT 0.5,
                context TEXT,
                PRIMARY KEY (page_id, entity_id),
                FOREIGN KEY (page_id) REFERENCES pages(id),
                FOREIGN KEY (entity_id) REFERENCES entities(id)
            )
        ''')
        
        # Notes/annotations
        cursor.execute('''
            CREATE TABLE IF NOT EXISTS notes (
                id TEXT PRIMARY KEY,
                page_id TEXT,
                content TEXT NOT NULL,
                created_at TIMESTAMP,
                tags TEXT,
                FOREIGN KEY (page_id) REFERENCES pages(id)
            )
        ''')
        
        # Create indexes
        cursor.execute('CREATE INDEX IF NOT EXISTS idx_pages_url ON pages(url)')
        cursor.execute('CREATE INDEX IF NOT EXISTS idx_pages_domain ON pages(domain)')
        cursor.execute('CREATE INDEX IF NOT EXISTS idx_pages_last_visit ON pages(last_visit)')
        cursor.execute('CREATE INDEX IF NOT EXISTS idx_entities_name ON entities(name)')
        cursor.execute('CREATE INDEX IF NOT EXISTS idx_entities_type ON entities(type)')
        
        conn.commit()
        conn.close()
    
    def _init_chromadb(self):
        """Initialize ChromaDB for vector search"""
        try:
            chroma_path = os.path.join(settings.memory_db_path, "chroma")
            self.vector_db = chromadb.PersistentClient(
                path=chroma_path,
                settings=ChromaSettings(anonymized_telemetry=False)
            )
            self.collection = self.vector_db.get_or_create_collection(
                name="pages",
                metadata={"hnsw:space": "cosine"}
            )
            print(f"[Memory] ChromaDB initialized with {self.collection.count()} documents")
        except Exception as e:
            print(f"[Memory] ChromaDB initialization failed: {e}")
            self.vector_db = None
            self.collection = None
    
    def _generate_id(self, content: str) -> str:
        """Generate a unique ID based on content"""
        return hashlib.sha256(content.encode()).hexdigest()[:16]
    
    async def remember_page(
        self,
        url: str,
        title: str,
        content: str,
        summary: Optional[str] = None,
        tags: List[str] = None,
        metadata: Dict[str, Any] = None
    ) -> Dict[str, Any]:
        """Store a page in memory"""
        
        from urllib.parse import urlparse
        
        page_id = self._generate_id(url)
        domain = urlparse(url).netloc
        now = datetime.now().isoformat()
        
        conn = sqlite3.connect(self.db_path)
        cursor = conn.cursor()
        
        # Check if page exists
        cursor.execute('SELECT id, visit_count FROM pages WHERE url = ?', (url,))
        existing = cursor.fetchone()
        
        if existing:
            # Update existing page
            cursor.execute('''
                UPDATE pages SET
                    title = ?,
                    content = ?,
                    summary = ?,
                    last_visit = ?,
                    visit_count = visit_count + 1,
                    tags = ?,
                    metadata = ?
                WHERE url = ?
            ''', (
                title,
                content[:50000],  # Limit content size
                summary,
                now,
                json.dumps(tags or []),
                json.dumps(metadata or {}),
                url
            ))
            page_id = existing[0]
        else:
            # Insert new page
            cursor.execute('''
                INSERT INTO pages (id, url, title, domain, content, summary, first_visit, last_visit, tags, metadata)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
            ''', (
                page_id,
                url,
                title,
                domain,
                content[:50000],
                summary,
                now,
                now,
                json.dumps(tags or []),
                json.dumps(metadata or {})
            ))
        
        conn.commit()
        conn.close()
        
        # Add to vector store for semantic search
        if self.collection:
            try:
                # Create a searchable text
                search_text = f"{title}\n{summary or ''}\n{content[:2000]}"
                
                self.collection.upsert(
                    ids=[page_id],
                    documents=[search_text],
                    metadatas=[{
                        "url": url,
                        "title": title,
                        "domain": domain,
                        "timestamp": now
                    }]
                )
            except Exception as e:
                print(f"[Memory] Vector store error: {e}")
        
        return {
            "id": page_id,
            "url": url,
            "title": title,
            "stored": True
        }
    
    async def search(
        self,
        query: str,
        limit: int = 10,
        domain: Optional[str] = None,
        min_date: Optional[str] = None
    ) -> List[Dict[str, Any]]:
        """Search memory using semantic similarity"""
        
        results = []
        
        # Vector search if available
        if self.collection and self.collection.count() > 0:
            try:
                vector_results = self.collection.query(
                    query_texts=[query],
                    n_results=min(limit, self.collection.count())
                )
                
                for i, doc_id in enumerate(vector_results['ids'][0]):
                    metadata = vector_results['metadatas'][0][i]
                    results.append({
                        "id": doc_id,
                        "url": metadata.get("url"),
                        "title": metadata.get("title"),
                        "domain": metadata.get("domain"),
                        "score": 1 - (vector_results['distances'][0][i] if vector_results.get('distances') else 0),
                        "source": "vector"
                    })
            except Exception as e:
                print(f"[Memory] Vector search error: {e}")
        
        # Fallback/supplement with SQLite full-text search
        conn = sqlite3.connect(self.db_path)
        cursor = conn.cursor()
        
        # Simple keyword search
        keywords = query.lower().split()
        like_clauses = " OR ".join([f"(title LIKE ? OR content LIKE ?)" for _ in keywords])
        params = []
        for kw in keywords:
            params.extend([f"%{kw}%", f"%{kw}%"])
        
        sql = f'''
            SELECT id, url, title, domain, summary, last_visit, visit_count
            FROM pages
            WHERE {like_clauses}
        '''
        
        if domain:
            sql += " AND domain = ?"
            params.append(domain)
        
        if min_date:
            sql += " AND last_visit >= ?"
            params.append(min_date)
        
        sql += " ORDER BY visit_count DESC, last_visit DESC LIMIT ?"
        params.append(limit)
        
        cursor.execute(sql, params)
        
        for row in cursor.fetchall():
            # Check if already in results
            if not any(r['id'] == row[0] for r in results):
                results.append({
                    "id": row[0],
                    "url": row[1],
                    "title": row[2],
                    "domain": row[3],
                    "summary": row[4],
                    "last_visit": row[5],
                    "visit_count": row[6],
                    "source": "keyword"
                })
        
        conn.close()
        
        return results[:limit]
    
    async def get_page(self, url: str) -> Optional[Dict[str, Any]]:
        """Get a specific page from memory"""
        conn = sqlite3.connect(self.db_path)
        cursor = conn.cursor()
        
        cursor.execute('''
            SELECT id, url, title, domain, content, summary, first_visit, last_visit, visit_count, tags, metadata
            FROM pages WHERE url = ?
        ''', (url,))
        
        row = cursor.fetchone()
        conn.close()
        
        if row:
            return {
                "id": row[0],
                "url": row[1],
                "title": row[2],
                "domain": row[3],
                "content": row[4],
                "summary": row[5],
                "first_visit": row[6],
                "last_visit": row[7],
                "visit_count": row[8],
                "tags": json.loads(row[9] or "[]"),
                "metadata": json.loads(row[10] or "{}")
            }
        return None
    
    async def get_recent(self, limit: int = 20) -> List[Dict[str, Any]]:
        """Get recently visited pages"""
        conn = sqlite3.connect(self.db_path)
        cursor = conn.cursor()
        
        cursor.execute('''
            SELECT id, url, title, domain, summary, last_visit, visit_count
            FROM pages
            ORDER BY last_visit DESC
            LIMIT ?
        ''', (limit,))
        
        results = []
        for row in cursor.fetchall():
            results.append({
                "id": row[0],
                "url": row[1],
                "title": row[2],
                "domain": row[3],
                "summary": row[4],
                "last_visit": row[5],
                "visit_count": row[6]
            })
        
        conn.close()
        return results
    
    async def get_related(self, url: str, limit: int = 5) -> List[Dict[str, Any]]:
        """Get pages related to a given URL"""
        
        # Get the page content for similarity search
        page = await self.get_page(url)
        if not page:
            return []
        
        # Use vector search to find similar pages
        if self.collection:
            try:
                search_text = f"{page['title']}\n{page.get('summary', '')}"
                results = self.collection.query(
                    query_texts=[search_text],
                    n_results=limit + 1,  # +1 to exclude self
                    where={"url": {"$ne": url}}
                )
                
                related = []
                for i, doc_id in enumerate(results['ids'][0]):
                    if results['metadatas'][0][i].get("url") != url:
                        related.append({
                            "id": doc_id,
                            "url": results['metadatas'][0][i].get("url"),
                            "title": results['metadatas'][0][i].get("title"),
                            "similarity": 1 - (results['distances'][0][i] if results.get('distances') else 0)
                        })
                
                return related[:limit]
            except Exception as e:
                print(f"[Memory] Related search error: {e}")
        
        # Fallback: same domain
        conn = sqlite3.connect(self.db_path)
        cursor = conn.cursor()
        
        cursor.execute('''
            SELECT id, url, title FROM pages
            WHERE domain = ? AND url != ?
            ORDER BY visit_count DESC
            LIMIT ?
        ''', (page['domain'], url, limit))
        
        results = [{"id": r[0], "url": r[1], "title": r[2]} for r in cursor.fetchall()]
        conn.close()
        
        return results
    
    async def add_note(
        self,
        page_url: str,
        content: str,
        tags: List[str] = None
    ) -> Dict[str, Any]:
        """Add a note to a page"""
        
        # Get page ID
        page = await self.get_page(page_url)
        if not page:
            return {"error": "Page not found in memory"}
        
        note_id = self._generate_id(f"{page_url}:{content}:{datetime.now().isoformat()}")
        
        conn = sqlite3.connect(self.db_path)
        cursor = conn.cursor()
        
        cursor.execute('''
            INSERT INTO notes (id, page_id, content, created_at, tags)
            VALUES (?, ?, ?, ?, ?)
        ''', (
            note_id,
            page['id'],
            content,
            datetime.now().isoformat(),
            json.dumps(tags or [])
        ))
        
        conn.commit()
        conn.close()
        
        return {
            "id": note_id,
            "page_id": page['id'],
            "content": content,
            "created": True
        }
    
    async def get_stats(self) -> Dict[str, Any]:
        """Get memory statistics"""
        conn = sqlite3.connect(self.db_path)
        cursor = conn.cursor()
        
        stats = {}
        
        cursor.execute('SELECT COUNT(*) FROM pages')
        stats['total_pages'] = cursor.fetchone()[0]
        
        cursor.execute('SELECT COUNT(DISTINCT domain) FROM pages')
        stats['unique_domains'] = cursor.fetchone()[0]
        
        cursor.execute('SELECT COUNT(*) FROM entities')
        stats['total_entities'] = cursor.fetchone()[0]
        
        cursor.execute('SELECT COUNT(*) FROM notes')
        stats['total_notes'] = cursor.fetchone()[0]
        
        cursor.execute('SELECT SUM(visit_count) FROM pages')
        stats['total_visits'] = cursor.fetchone()[0] or 0
        
        if self.collection:
            stats['vector_documents'] = self.collection.count()
        
        conn.close()
        return stats
    
    async def delete_page(self, url: str) -> bool:
        """Delete a page from memory"""
        page = await self.get_page(url)
        if not page:
            return False
        
        conn = sqlite3.connect(self.db_path)
        cursor = conn.cursor()
        
        # Delete related data
        cursor.execute('DELETE FROM page_entities WHERE page_id = ?', (page['id'],))
        cursor.execute('DELETE FROM notes WHERE page_id = ?', (page['id'],))
        cursor.execute('DELETE FROM pages WHERE id = ?', (page['id'],))
        
        conn.commit()
        conn.close()
        
        # Remove from vector store
        if self.collection:
            try:
                self.collection.delete(ids=[page['id']])
            except:
                pass
        
        return True
    
    async def clear_all(self) -> bool:
        """Clear all memory data"""
        conn = sqlite3.connect(self.db_path)
        cursor = conn.cursor()
        
        cursor.execute('DELETE FROM page_entities')
        cursor.execute('DELETE FROM entities')
        cursor.execute('DELETE FROM notes')
        cursor.execute('DELETE FROM pages')
        
        conn.commit()
        conn.close()
        
        # Clear vector store
        if self.vector_db and self.collection:
            try:
                self.vector_db.delete_collection("pages")
                self.collection = self.vector_db.create_collection(
                    name="pages",
                    metadata={"hnsw:space": "cosine"}
                )
            except:
                pass
        
        return True


# Singleton instance
memory = MemoryStore()

File: src/ai/memory.js
// EVOS Browser - AI Memory System
// Simple vector-based memory using local embeddings

const fs = require('fs');
const path = require('path');
const { MEMORY_PATH, MEMORY_CONFIG, ensureDirectories } = require('./config');

class AIMemory {
  constructor() {
    this.memories = [];
    this.memoryFile = path.join(MEMORY_PATH, 'memories.json');
    this.embeddingModel = null;
    this.isInitialized = false;
    ensureDirectories();
  }

  // Initialize memory system
  async initialize() {
    if (this.isInitialized) return;
    
    try {
      // Load existing memories
      await this.loadMemories();
      
      this.isInitialized = true;
      console.log(`[Memory] Initialized with ${this.memories.length} memories`);
    } catch (error) {
      console.error('[Memory] Initialization error:', error);
    }
  }

  // Load memories from disk
  async loadMemories() {
    try {
      if (fs.existsSync(this.memoryFile)) {
        const data = fs.readFileSync(this.memoryFile, 'utf-8');
        this.memories = JSON.parse(data);
      }
    } catch (error) {
      console.error('[Memory] Failed to load memories:', error);
      this.memories = [];
    }
  }

  // Save memories to disk
  async saveMemories() {
    try {
      ensureDirectories();
      fs.writeFileSync(this.memoryFile, JSON.stringify(this.memories, null, 2));
    } catch (error) {
      console.error('[Memory] Failed to save memories:', error);
    }
  }

  // Simple text similarity using word overlap (TF-IDF like)
  calculateSimilarity(text1, text2) {
    const tokenize = (text) => {
      return text.toLowerCase()
        .replace(/[^\w\s]/g, '')
        .split(/\s+/)
        .filter(w => w.length > 2);
    };

    const words1 = new Set(tokenize(text1));
    const words2 = new Set(tokenize(text2));
    
    if (words1.size === 0 || words2.size === 0) return 0;

    const intersection = [...words1].filter(w => words2.has(w));
    const union = new Set([...words1, ...words2]);
    
    // Jaccard similarity
    return intersection.length / union.size;
  }

  // Add a memory
  async addMemory(content, metadata = {}) {
    const memory = {
      id: Date.now().toString(36) + Math.random().toString(36).substr(2),
      content: content,
      metadata: {
        ...metadata,
        timestamp: Date.now(),
        date: new Date().toISOString()
      }
    };

    // Check for duplicates (similar content)
    const isDuplicate = this.memories.some(m => 
      this.calculateSimilarity(m.content, content) > 0.8
    );

    if (!isDuplicate) {
      this.memories.push(memory);
      
      // Limit memory count
      if (this.memories.length > MEMORY_CONFIG.maxMemories) {
        // Remove oldest memories
        this.memories = this.memories.slice(-MEMORY_CONFIG.maxMemories);
      }
      
      await this.saveMemories();
      console.log('[Memory] Added new memory:', memory.id);
      return memory;
    } else {
      console.log('[Memory] Skipped duplicate memory');
      return null;
    }
  }

  // Search memories by similarity
  async search(query, limit = 5) {
    if (this.memories.length === 0) return [];

    // Calculate similarity scores
    const scored = this.memories.map(memory => ({
      memory,
      score: this.calculateSimilarity(query, memory.content)
    }));

    // Sort by score and filter
    const results = scored
      .filter(item => item.score >= MEMORY_CONFIG.similarityThreshold)
      .sort((a, b) => b.score - a.score)
      .slice(0, limit)
      .map(item => ({
        ...item.memory,
        relevanceScore: item.score
      }));

    return results;
  }

  // Full-text search
  async fullTextSearch(query, limit = 10) {
    const queryLower = query.toLowerCase();
    
    const results = this.memories
      .filter(m => m.content.toLowerCase().includes(queryLower))
      .slice(0, limit);
    
    return results;
  }

  // Get recent memories
  async getRecent(limit = 10) {
    return this.memories
      .sort((a, b) => b.metadata.timestamp - a.metadata.timestamp)
      .slice(0, limit);
  }

  // Get memories by type/tag
  async getByType(type, limit = 10) {
    return this.memories
      .filter(m => m.metadata.type === type)
      .slice(0, limit);
  }

  // Get memory by ID
  async getById(id) {
    return this.memories.find(m => m.id === id);
  }

  // Delete a memory
  async deleteMemory(id) {
    const index = this.memories.findIndex(m => m.id === id);
    if (index !== -1) {
      this.memories.splice(index, 1);
      await this.saveMemories();
      return true;
    }
    return false;
  }

  // Clear all memories
  async clearAll() {
    this.memories = [];
    await this.saveMemories();
  }

  // Get memory stats
  getStats() {
    const types = {};
    this.memories.forEach(m => {
      const type = m.metadata.type || 'unknown';
      types[type] = (types[type] || 0) + 1;
    });

    return {
      totalMemories: this.memories.length,
      byType: types,
      oldestMemory: this.memories[0]?.metadata.date,
      newestMemory: this.memories[this.memories.length - 1]?.metadata.date
    };
  }

  // Remember a browsing context
  async rememberPage(pageInfo) {
    const content = `
      Visited page: ${pageInfo.title}
      URL: ${pageInfo.url}
      ${pageInfo.description ? `Description: ${pageInfo.description}` : ''}
      ${pageInfo.summary ? `Summary: ${pageInfo.summary}` : ''}
    `.trim();

    return await this.addMemory(content, {
      type: 'page_visit',
      url: pageInfo.url,
      title: pageInfo.title
    });
  }

  // Remember a user interaction
  async rememberInteraction(action, result) {
    const content = `
      User action: ${action}
      Result: ${result}
    `.trim();

    return await this.addMemory(content, {
      type: 'interaction',
      action: action
    });
  }

  // Remember a fact or piece of information
  async rememberFact(fact, source = null) {
    return await this.addMemory(fact, {
      type: 'fact',
      source: source
    });
  }

  // Get context for the AI (combines recent + relevant memories)
  async getContext(currentQuery, maxTokens = 1000) {
    let context = '';
    
    // Get relevant memories
    const relevant = await this.search(currentQuery, 3);
    
    // Get recent memories
    const recent = await this.getRecent(3);
    
    // Combine and deduplicate
    const combined = new Map();
    [...relevant, ...recent].forEach(m => {
      if (!combined.has(m.id)) {
        combined.set(m.id, m);
      }
    });
    
    // Build context string
    const memories = Array.from(combined.values());
    
    if (memories.length > 0) {
      context = 'Relevant memories:\n';
      for (const memory of memories) {
        const memoryText = `- ${memory.content}\n`;
        if ((context + memoryText).length < maxTokens) {
          context += memoryText;
        }
      }
    }
    
    return context;
  }
}

// Singleton instance
const aiMemory = new AIMemory();

module.exports = { AIMemory, aiMemory };
